tests/BasicsTest.hack

```diff
@@ -68,6 +100,103 @@ class BasicsTest extends Facebook\HackTest\HackTest {
     );
   }
 
+  public function provideFaultyTrees(): vec<(x\node, string)> {
+    $br = <br />;
+    $ui_div = <ui:div><div><span /></div></ui:div>;
+    /*HHAST_IGNORE_ERROR[DontUseAsioJoin]*/
+    \HH\Asio\join($br->toStringAsync());
+    /*HHAST_IGNORE_ERROR[DontUseAsioJoin]*/
+    \HH\Asio\join($ui_div->toStringAsync());
+
+    return vec[
+      tuple(
+        <div><span>{$br}</span></div>,
+        // Normal path
+        'Via XHPPath: HTML\\div -> HTML\\span -> HTML\\br.',
+      ),
+      tuple(
+        <div><ui:div>{$br}</ui:div></div>,
+        // Works for custom elements
+        'Via XHPPath: HTML\\div -> ui\\div -> HTML\\div -> HTML\\br.',
+      ),
+      tuple(
+        <div><x:frag>{$br}</x:frag></div>,
+        // Frags are not included in the XHPPath
+        'Via XHPPath: HTML\\div -> HTML\\br.',
+      ),
+      tuple(
+        <div><span>{$ui_div}</span></div>,
+        // Custom element at the end of the chain
+        'Via XHPPath: HTML\\div -> HTML\\span -> ui\\div.',
+      ),
+      tuple(
+        <ui:div><span>{$ui_div}</span></ui:div>,
+        // Custom element at the start of the chain
+        'Via XHPPath: ui\\div -> HTML\\div -> HTML\\span -> ui\\div.',
+      ),
+      tuple(
+        <div>
+          <span />
+          <div><span /></div>
+          <ui:div><span>{$br}</span></ui:div>
+          <ui:div />
+        </div>,
+        // In a subtree
+        'Via XHPPath: HTML\\div -> ui\\div -> HTML\\div -> HTML\span -> HTML\\br.',
+      ),
+      tuple(
+        <div>
+          <ui:div><ui:div><span>{$br}</span></ui:div></ui:div>
+        </div>,
+        // Custom inside of custom
+        'Via XHPPath: HTML\\div -> ui\\div -> HTML\\div -> ui\\div -> HTML\\div -> HTML\span -> HTML\\br.',
+      ),
+      tuple(
+        <ui:returning:ui><span>{$br}</span></ui:returning:ui>,
+        // Limitation a ui:div is missing V here, because Element::__flushRenderedRootElement loops until a primitive
+        'Via XHPPath: ui\\returning\\ui -> HTML\\div -> HTML\span -> HTML\\br.',
+      ),
```

Comment:
This is an important limitation keep in mind. Building the path in the way that we are doing has no extra runtime cost, compared to the code in master. You don't pay for creating the path, unless an exception is thrown. And when the exception is thrown, the perf argument doesn't matter much, since you failed to create a valid XHP tree.

However [Element::__flushRenderedRootElement()](https://github.com/hhvm/xhp-lib/blob/b570b9d34d3b457d1a3a40135c85934d81ab5749/src/core/Element.hack#L85-L100) will loop in place and lose the names of the intermediary custom classes. Adding a vec here that collects the classnames as they come by is trivial, but then you are paying a runtime cost for collecting those names, even though you are (in 99.99% of the cases) going to throw this vec away without looking at it again.

I am imagining this loop to be hot code at most big sites and paying a perf cost here for a debugging feature feels destructive. What do you think would be the best option?

 - Don't collect an XHPPath at all, only include the name of the element that threw in the exception message.
 - Only collect in Primitive (this means that you'll only see the primitives in an XHPPath).
 - Collect the vec if a flag is set (similar to how ChildValidation can be enabled an disabled).
 - Collect the vec, XHP is already doing far more CPU intensive work by rendering the tree.

Reply:
The current solution (keeping the vec in the exception object) seems good to me, even if it might not fully handle all possible cases. Including only the name of the doubly-rendered element seems fine to me too.